sql/database.go

```diff
@@ -202,64 +225,79 @@ func InMemory(opts ...Opt) *sqliteDatabase {
 // https://www.sqlite.org/pragma.html#pragma_synchronous
 func Open(uri string, opts ...Opt) (*sqliteDatabase, error) {
 	config := defaultConf()
+	config.uri = uri
 	for _, opt := range opts {
 		opt(config)
 	}
-	logger := config.logger.With(zap.String("uri", uri))
+	if !config.temp && config.handleIncompleteMigrations {
+		if err := handleIncompleteCopyMigration(config); err != nil {
+			return nil, err
+		}
+	}
+	return openDB(config)
+}
+
+func openDB(config *conf) (*sqliteDatabase, error) {
+	logger := config.logger.With(zap.String("uri", config.uri))
 	var flags sqlite.OpenFlags
 	if !config.forceFresh {
 		flags = sqlite.SQLITE_OPEN_READWRITE |
-			sqlite.SQLITE_OPEN_WAL |
 			sqlite.SQLITE_OPEN_URI |
 			sqlite.SQLITE_OPEN_NOMUTEX
+		if !config.temp {
+			// Note that SQLITE_OPEN_WAL is not handled by SQLITE api itself,
+			// but rather by the crawshaw library which executes
+			// PRAGMA journal_mode=WAL in this case.
+			// We don't want it for temporary databases as they're not
+			// using any journal
+			flags |= sqlite.SQLITE_OPEN_WAL
+		}
 	}
 	freshDB := config.forceFresh
-	pool, err := sqlitex.Open(uri, flags, config.connections)
+	pool, err := sqlitex.Open(config.uri, flags, config.connections)
 	if err != nil {
 		if config.forceFresh || sqlite.ErrCode(err) != sqlite.SQLITE_CANTOPEN {
-			return nil, fmt.Errorf("open db %s: %w", uri, err)
+			return nil, fmt.Errorf("open db %s: %w", config.uri, err)
 		}
 		flags |= sqlite.SQLITE_OPEN_CREATE
 		freshDB = true
-		pool, err = sqlitex.Open(uri, flags, config.connections)
+		pool, err = sqlitex.Open(config.uri, flags, config.connections)
 		if err != nil {
-			return nil, fmt.Errorf("create db %s: %w", uri, err)
+			return nil, fmt.Errorf("create db %s: %w", config.uri, err)
 		}
 	}
 	db := &sqliteDatabase{pool: pool}
 	if config.enableLatency {
 		db.latency = newQueryLatency()
 	}
+
+	if config.temp {
+		// Temporary database is used for migration and is deleted if migrations
+		// fail, so we make it faster by disabling journaling and synchronous
+		// writes.
+		if _, err := db.Exec("PRAGMA journal_mode=OFF", nil, nil); err != nil {
+			return nil, errors.Join(
+				fmt.Errorf("PRAGMA journal_mode=OFF: %w", err),
+				db.Close())
+		}
+		if _, err := db.Exec("PRAGMA synchronous=OFF", nil, nil); err != nil {
+			return nil, errors.Join(
+				fmt.Errorf("PRAGMA journal_mode=OFF: %w", err),
+				db.Close())
```

Comment:
I don't think this adds any meaningful additional info. If lets say PRAGMA failed because of an SQLITE_IOERR the returned error will be

```
PARAGMA joural_mode=OFF: prepare PRAMGA synchronous=OFF: SQL_IOERR Con.Prepare\nclose pool SQL_IOERR Conn.Close
```

(btw. the `fmt.Errof` text is incorrect)
This doesn't add any additional information, the error happened on the db.Exec - which likely will also cause `db.Close()` to fail (but not necessarily) - and if it does it will fail most likely with exactly the same error again?

Reply:
My point is if we establish something like this as general pattern, the error we get at the top level will be a wall of text of mostly the same error (or an error by the same cause) repeated multiple times...

Reply:
Looking [at the SQLite source](https://github.com/sqlite/sqlite/blob/75bdf00e8980610223acf865245b7052292a1102/src/main.c#L1261-L1317), `SQLITE_IOERR` will never happen during `db.Close()`. In the absolute majority of cases, `db.Close()` will actually succeed even if the DB is broken. Yet it may return `SQLITE_MISUSE_BKPT` or `SQLITE_BUSY`, which may be quite different from what `db.Exec` has returned, if there's something horribly wrong with our database connection handling, which better shouldn't go unnoticed

Reply:
I agree with @fasmat here. I don't think that `db.Close` error should be returned here too. It is a best effort `Close` and it isn't clear what purpose does the error serve. I would swallow the `Close` and report only the error from `Exec`

Reply:
Removed `errors.Join`